### PR TITLE
feat(post): add prescaling post

### PR DESCRIPTION
**Pull request by taufort**
Blog post aiming at presenting the prescaling solution we set up on Bedrock's Kubernetes clusters to handle sudden and abrupt, but predictable traffic spikes :slightly_smiling_face:

### CODE REVIEW

File: _posts/2022-02-03-prescaling.md [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+    name: Timothée Aufort
+avatar:
+email:
+twitter: TimAufort
+github: taufort
+category:
+tags: [kubernetes, scaling, high availability, aws, cloud]
+comments: false
+image:
+  feature: <IMAGE>
```
Comment (taufort) [inline _posts/2022-02-03-prescaling.md:15]: Find the main image for the article

File: _posts/2022-02-03-prescaling.md [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+We know our applications receive a sudden and brutal traffic spike once a day, between 20:50 and 21:00 Paris time. 
+Not taking special events into account, that’s the only time load increases violently enough for reactive autoscaling 
+to be unable to handle it – and, most days, it actually does quite fine. So, we could pre-provision more capacity 
+around that time (only), and not pay for it the rest of the day…
+
+## Prescaling our applications
+
+### Enabling and configuring prescaling on an HPA
+
+To enable and configure prescaling on an HorizontalPodAutoscaler, you must add two sets of information:
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:179]:
```suggestion
To enable and configure prescaling on an `HorizontalPodAutoscaler`, we add two sets of information:
```

File: _posts/2022-02-03-prescaling.md [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+* Three annotations to define:
+  * Time when prescaling starts.
+  * Time when prescaling stops.
+  * The minimum number of Pods you want between those two times.
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:183]:
```suggestion
  * The minimum number of Pods we want between those two times.
```

File: _posts/2022-02-03-prescaling.md [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+    # …
+```
+
+In this example, we indicate we want at least 25 pods between 19:30:00 and 23:30:00. As a result, 
+the number of pods of this application will rise up to at least 25 pods during this period of time.
+
+Times are expressed in the local timezone of the Kubernetes cluster as prescaling is linked to events on the platform. 
+Those events are usually linked to events on live TV and we deploy one cluster per TV broadcaster.
+
+In the metrics block, you need to configure a new external metric:
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:206]:
```suggestion
In the `metrics` block, you need to configure a new `External` metric:
```

File: _posts/2022-02-03-prescaling.md [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+ - type: External
+   external:
+     metricName: "annotation_scaling_min_replica"
+     metricSelector:
+       matchLabels:
+         deployment: "service-6play-images"
+     targetValue: "10"
+```
+
+The label used for `deployment` must be set to the value of `scaleTargetRef.name` (= the name of the Deployment the
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:220]:
```suggestion
The label used for `deployment` must be set to the value of `scaleTargetRef.name` (= the name of the `Deployment` the 
```

File: _posts/2022-02-03-prescaling.md [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+   external:
+     metricName: "annotation_scaling_min_replica"
+     metricSelector:
+       matchLabels:
+         deployment: "service-6play-images"
+     targetValue: "10"
+```
+
+The label used for `deployment` must be set to the value of `scaleTargetRef.name` (= the name of the Deployment the 
+HPA reconfigures). The `targetValue` must always be set to 10.
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:221]:
```suggestion
HPA reconfigures). The `targetValue` must always be set to `10`.
```

File: _posts/2022-02-03-prescaling.md [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+
+You now know how we configure an HPA for prescaling. What you do not know yet is how the HPA annotations are used 
+and which application exposes the metrics called `annotation_scaling_min_replica`. It’s now time to talk about the 
+prescaling exporter.
+
+### The prescaling exporter
+
+The prescaling exporter is a [Prometheus](https://prometheus.io/) exporter developed in Python.
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:229]:
```suggestion
The prescaling exporter is a [Prometheus](https://prometheus.io/) exporter we developed (in Python). 
```

File: _posts/2022-02-03-prescaling.md [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+
+> Prometheus is one of the tools we use in our monitoring stack at Bedrock. One of its purposes, among others, is to 
+collect metrics from Pods. This article will not present our Prometheus stack in detail. 
+
+Here is how the prescaling exporter works:
+
+![Prescaling v1](/images/posts/2022-02-03-prescaling/prescaling_v1.png)
+
+1) Every 15 seconds or so, Prometheus scrapes the prescaling exporter pod to get the metrics it exposes.
+2) Scrapping triggers the generation of the Prometheus metrics. Before the metrics are exposed, the exporter first
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:240]:
```suggestion
2) Scrapping triggers the generation of the metrics. Before they are exposed, the exporter first 
```

File: _posts/2022-02-03-prescaling.md [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+
+How does the prescaling exporter calculate the metrics of the HPA subscribed to the prescaling? Well, it depends 
+on the content of the annotations you configured on the HPA.
+
+Here is how a Prometheus metric of the prescaling exporter looks like:
+
+![Prometheus metric](/images/posts/2022-02-03-prescaling/prometheus_metric.png)
+
+In each metric, you will find several labels. I chose to put only one here to simplify.
+The metric can take three possible values depending on the content of the annotations we saw earlier:
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:256]:
```suggestion
The metric can take one of three values, depending on the content of the annotations we saw earlier:
```

File: _posts/2022-02-03-prescaling.md [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+* When we are not within the time range of the prescaling annotations of the HPA, it means that we do not 
+need to prescale. As a result, the metric is set to 0.
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:258]:
```suggestion
need to prescale. As a result, the metric is set to `0`.
```

File: _posts/2022-02-03-prescaling.md [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+* If we are within the time range of the prescaling annotations, it means that we are in the prescaling time range.
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:259]:
```suggestion
* If we are within the time range of the prescaling annotations, it means we are in the prescaling time range. 
```

File: _posts/2022-02-03-prescaling.md [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+![Prometheus metric](/images/posts/2022-02-03-prescaling/prometheus_metric.png)
+
+In each metric, you will find several labels. I chose to put only one here to simplify.
+The metric can take three possible values depending on the content of the annotations we saw earlier:
+* When we are not within the time range of the prescaling annotations of the HPA, it means that we do not 
+need to prescale. As a result, the metric is set to 0.
+* If we are within the time range of the prescaling annotations, it means that we are in the prescaling time range. 
+From there, two possibilities:
+  * If “Current number of replicas” < “Number of minimum replicas in the HPA annotation”, we need to add replicas so 
+  the metric is set to 11.
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:262]:
```suggestion
  the metric is set to `11`.
```

File: _posts/2022-02-03-prescaling.md [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+  * If “Current number of replicas” >= “Number of minimum replicas in the HPA annotation”, we have enough replicas so 
+  the metric is set to 10.
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:264]:
```suggestion
  the metric is set to `10`.
```

File: _posts/2022-02-03-prescaling.md [hunk 14]
```diff
@@ -0,0 +1,376 @@
...
+
+> When the metric is set to 10 and we already have enough replicas running, the number of replicas will never go below 
+the minimum chosen in the prescaling annotation `annotations.scaling.exporter.replica.min`.
+
+Here is what it looks like on Grafana for the application `service-6play-images` during the evening:
+
+![annotation_scaling_min_replica over time](/images/posts/2022-02-03-prescaling/annotation_scaling_min_replica_over_time.png)
+<center><i>annotation_scaling_min_replica over time</i></center>
+
+In this example, we can see that:
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:274]:
```suggestion
In this example:
```

File: _posts/2022-02-03-prescaling.md [hunk 15]
```diff
@@ -0,0 +1,376 @@
...
+
+### How can the HPAs prescale?
+
+To understand how the HPAs can prescale, we need to talk about 
+[the prometheus adapter](https://github.com/kubernetes-sigs/prometheus-adapter). It is an implementation 
+of the Kubernetes metrics APIs. We use it to expose custom and external metrics for HPAs to use in order to scale:
+
+![HPA scale](/images/posts/2022-02-03-prescaling/hpa_scale.png)
+
+1) Prometheus adapter collects metrics from prometheus once every minute and exposes them as external and
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:292]:
```suggestion
1) Prometheus adapter collects metrics from prometheus once every minute and exposes them as `External` and 
```

File: _posts/2022-02-03-prescaling.md [hunk 16]
```diff
@@ -0,0 +1,376 @@
...
+### How can the HPAs prescale?
+
+To understand how the HPAs can prescale, we need to talk about 
+[the prometheus adapter](https://github.com/kubernetes-sigs/prometheus-adapter). It is an implementation 
+of the Kubernetes metrics APIs. We use it to expose custom and external metrics for HPAs to use in order to scale:
+
+![HPA scale](/images/posts/2022-02-03-prescaling/hpa_scale.png)
+
+1) Prometheus adapter collects metrics from prometheus once every minute and exposes them as external and 
+custom metrics.
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:293]:
```suggestion
`Custom` metrics.
```

File: _posts/2022-02-03-prescaling.md [hunk 17]
```diff
@@ -0,0 +1,376 @@
...
+we’ve seen before, for all HPAs bearing this annotation”.
+
+To deal with those very special events, we added another component in the prescaling stack: the prescaling API.
+
+### The prescaling API
+
+The prescaling API is a backend application also developed in Python. It was designed to store prescaling 
+settings for future events on the platform in AWS DynamoDB. We chose DynamoDB because it’s a serverless database 
+easily maintainable through Terraform code. By “event”, understand a football game or another big show such as 
+Top Chef or whatnot. Those settings  define when and how we must enlarge the platform to sustain bigger
```
Comment (pmartin) [inline _posts/2022-02-03-prescaling.md:335]:
```suggestion
Top Chef. Those settings define when and how we must enlarge the platform to sustain bigger 
```

File: _posts/2022-02-03-prescaling.md [hunk 18]
```diff
@@ -0,0 +1,376 @@
...
+image:
+  feature: posts/bonnes-pratiques-web/bedrock.jpg
+language: en
+---
+
+Are you experiencing peak loads on your Kubernetes-hosted platform? Rest assured, you are not alone.
+At Bedrock, we have developed a prescaling solution. It allows us to handle sudden and abrupt, but predictable, 
+traffic spikes, like soccer games. Kubernetes provides HorizontalPodAutoscalers to handle traffic variations. 
+We'll look at their limitations in the case of meteoric spikes in load and how prescaling helps us deal with the 
+sudden arrival of several hundred thousand users.
```
[the remark this answers is not in the document]
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:23]: J'ai fait des retours chariots simples pour que ce soit plus facilement lisible dans un IDE :) 

J'ai ajouté un saut de ligne pour l'intro juste après `like soccer games`

File: _posts/2022-02-03-prescaling.md [hunk 19]
```diff
@@ -0,0 +1,376 @@
...
+
+To deal with these load variations, several tools help us to automatically adapt our Kubernetes  clusters’ capacity:
+* [HorizontalPodAutoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) 
+(HPA): adds/removes Pods (= capacity) on a workload resource such as Deployment or a 
+StatefulSet.
+* [Cluster Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler): 
+adjusts the size of a Kubernetes cluster by adding/removing nodes.
+* [Overprovisioning](https://github.com/kubernetes-sigs/cluster-proportional-autoscaler): starts “empty” 
+pods (and new “useless” nodes, as a consequence), so the cluster has available capacity that will be used to start 
+applications pods quicker.
```
[the remark this answers is not in the document]
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:59]: Both are fine in my opinion. They are synonyms.

File: _posts/2022-02-03-prescaling.md [hunk 20]
```diff
@@ -0,0 +1,376 @@
...
+
+## Beginning of the scaling problems
+
+On an ordinary evening, when the multiple Kubernetes scaling tools were kicking in, here is basically what was 
+happening:
+
+![Traffic normal evening](/images/posts/2022-02-03-prescaling/traffic_normal_evening.png)
+
+As the load increased, our capacity was increasing as well and we always had spare capacity. We were able to double our 
+initial capacity every 5 minutes thanks to reactive scaling when load started to arise.
```
[the remark this answers is not in the document]
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:77]: Good catch. Fixed!

File: _posts/2022-02-03-prescaling.md [hunk 21]
```diff
@@ -0,0 +1,376 @@
...
+
+After a while and on some special evenings, we began to see this kind of behavior:
+
+![Traffic special evening](/images/posts/2022-02-03-prescaling/traffic_special_evening.png)
+
+Sometimes, the load was increasing more than the reactive scaling could handle, that is to say about x2 in capacity
```
[the remark this answers is not in the document]
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:83]: Fixed

File: _posts/2022-02-03-prescaling.md [hunk 22]
```diff
@@ -0,0 +1,376 @@
...
+      name: cpu
+      targetAverageUtilization: 70
+  - type: Pods
+    pods:
+      metricName: phpfpm_active_process_percentage
+      targetAverageValue: "40"
+  # …
+```
+
+In here, you should notice:
```
[the remark this answers is not in the document]
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:121]: Updated

File: _posts/2022-02-03-prescaling.md [hunk 23]
```diff
@@ -0,0 +1,376 @@
...
+* `minReplicas` is the minimum number of Pods that will run (under normal conditions).
+* `maxReplicas` is the maximum number of Pods that will run.
+* `metrics` is a list of metrics the HPA will analyze to determine if it must scale or not. If usage gets higher 
+than the target specified for a metric, the HPA will add new Pods (= scale out). If usage gets lower than all targets 
+specified for all metrics, the HPA will remove Pods (= scale in).
+
+### What scale out looks like in a real case
+
+At Bedrock, we use 3 types of metrics to scale: Resource, Custom and External. Here is an example of an application
```
[the remark this answers is not in the document]
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:130]: Fixed

File: _posts/2022-02-03-prescaling.md [hunk 24]
```diff
@@ -0,0 +1,376 @@
...
+
+### What scale out looks like in a real case
+
+At Bedrock, we use 3 types of metrics to scale: Resource, Custom and External. Here is an example of an application 
+scaling out on a Custom metric:
+
+![Reactive scaling process percentage](/images/posts/2022-02-03-prescaling/reactive_scaling_process_percentage.png)
+<center><i>Process status (%) over time</i></center>
+
+In this graph, around 20:52:30 and 20:55:00, we can see that the percentage of active processes rises to go above
```
[the remark this answers is not in the document]
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:136]: Fixed

File: _posts/2022-02-03-prescaling.md [hunk 25]
```diff
@@ -0,0 +1,376 @@
...
+![Reactive scaling process percentage](/images/posts/2022-02-03-prescaling/reactive_scaling_process_percentage.png)
+<center><i>Process status (%) over time</i></center>
+
+In this graph, around 20:52:30 and 20:55:00, we can see that the percentage of active processes rises to go above 
+the target 40 we saw in the YAML example before. This triggers the scale out of the HPA of this application:
+
+![Reactive scaling pods](/images/posts/2022-02-03-prescaling/reactive_scaling_pods.png)
+<center><i>Pod status over time</i></center>
+
+We can see that around 20:53:00 (30 seconds after we first go above the target), the number of pods unavailable rises.
```
[the remark this answers is not in the document]
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:142]: Done

File: _posts/2022-02-03-prescaling.md [hunk 26]
```diff
@@ -0,0 +1,376 @@
...
+* We might not have enough EC2 servers running to host the new Pods that are trying to start (we usually have between 
+10% and 20% of spare capacity). If starting new EC2 servers is necessary, this takes between 2 and 4 minutes.
+* Pods (as our applications and Docker images are big) usually need between 30 and 60 seconds to start.
+
+This means when a spike in load occurs, we need between 1’00’’ and 6’00’’ for new Pods to be able to handle that load. 
+Scaling is clearly not instantaneous.
+
+## Prescaling… What is this about?
+
+Most of our applications deployed in Kubernetes use an HorizontalPodAutoscaler. As we’ve seen, this autoscaling
```
[the remark this answers is not in the document]
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:162]: Both are proper English but why not.

File: _posts/2022-02-03-prescaling.md (hunk 2 above)
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:181]: Done

File: _posts/2022-02-03-prescaling.md (hunk 3 above)
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:185]: Fixed

File: _posts/2022-02-03-prescaling.md (hunk 4 above)
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:208]: Done

File: _posts/2022-02-03-prescaling.md (hunk 5 above)
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:222]: Fixed

File: _posts/2022-02-03-prescaling.md (hunk 6 above)
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:223]: Fixed

File: _posts/2022-02-03-prescaling.md (hunk 8 above)
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:242]: Fair enough

File: _posts/2022-02-03-prescaling.md (hunk 10 above)
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:260]: Fixed

File: _posts/2022-02-03-prescaling.md (hunk 9 above)
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:258]: Fixed

File: _posts/2022-02-03-prescaling.md (hunk 11 above)
Comment (taufort) [reply _posts/2022-02-03-prescaling.md:261]: Better indeed

File: _posts/2022-02-03-prescaling.md (hunk 1 above)
Comment (taufort) [reply _posts/2022-02-03-prescaling.md]: Done!

**Pull request merged**